src/types.jl

```diff
@@ -189,10 +224,15 @@ for (type, base_type, _) in ABSTRACT_QUANTITY_TYPES
     end
 end
 
-const DEFAULT_QUANTITY_TYPE = Quantity{DEFAULT_VALUE_TYPE, DEFAULT_DIM_TYPE}
+const DEFAULT_QUANTITY_TYPE = RealQuantity{DEFAULT_VALUE_TYPE, DEFAULT_DIM_TYPE}
 
-new_dimensions(::Type{D}, dims...) where {D<:AbstractDimensions} = constructorof(D)(dims...)
-new_quantity(::Type{Q}, l, r) where {Q<:UnionAbstractQuantity} = constructorof(Q)(l, r)
+@inline function new_dimensions(::Type{D}, dims...) where {D<:AbstractDimensions}
+    return constructorof(D)(dims...)
+end
+@inline function new_quantity(::Type{Q}, val, dims) where {Q<:UnionAbstractQuantity}
+    Qout = promote_quantity(Q, typeof(val))
+    return constructorof(Qout)(val, dims)
+end
```

Comment:
Now `new_quantity` will automatically promote from `RealQuantity -> Quantity` when needed. (Just like how the value type can also change.)

This hierarchy is defined by `promote_quantity(::Type{<:UnionAbstractQuantity}, ::Type{<:Any})` which is given for the default concrete types, but which the user can obviously overload.